### ISSUE TITLE
Vulnerability in Amcharts 5

### ISSUE THREAD

**Issue by jlgonzalezmindden** [amcharts/amcharts5#897]
Hi,

I am using the latest version of Amcharts 5 specifically the version 5.3.11 recently a vulnerability has appeared with a dependency that they are using xlsx. 

Link to vulnerability: https://github.com/advisories/GHSA-4r6h-8v6p-xvw6

Thanks in advance.

**Comment by martynasma** [comment]
OK, so it looks like SheetJS team had some [weird beef](https://git.sheetjs.com/sheetjs/sheetjs/issues/2667) with NPM and stopped publishing their packages there since version 0.18.5.

So there's really no way for us to include that as a dependency, save for completely forking their repo, and including that into amCharts. That is an option that brings significant maintenance overheads, so we'd like to avoid that.

Given, that none of the amCharts functionality uses parts of xlsx affected by the vulnerability, we're going to keep using current version, unless a) they have a change of heart and get back on NPM; b) or a better alternative to exporting XLSX comes around; c) or this somehow starts affecting amCharts functionality.

Hope that sounds reasonable.

**Comment by SymbioticKilla** [comment]
@martynasma 
Hi,

thank you for clarification.
Just for information:
Your package ist used in productive code. There are companies with 0 CVE policy. Please consider to find the solution as soon as possible.

Thanks!

**Comment by dasa** [comment]
> So there's really no way for us to include that as a dependency

Would their suggestion [here](https://docs.sheetjs.com/docs/getting-started/installation/frameworks) work?

```console
npm rm --save xlsx
npm i --save https://cdn.sheetjs.com/xlsx-0.19.3/xlsx-0.19.3.tgz
```

**Comment by dasa** [comment]
Or for Yarn:

```console
yarn remove xlsx
yarn add https://cdn.sheetjs.com/xlsx-0.19.3/xlsx-0.19.3.tgz
```

**Comment by Pauan** [comment]
@SymbioticKilla You will have to ask XLSX to fix it, because it is their fault, not ours. 

----

@dasa That would cause all of our users to use the XLSX CDN, which we will not do. There are a lot of issues with using a CDN as a dependency, including security issues.

CDNs cannot be auto-upgraded, they cannot have security audits, they do not have checksums, they are not verified by npm, they are vulnerable to various security attacks, etc.

Instead, you can add this into your `package.json`:

```json
"resolutions": {
  "xlsx": "https://cdn.sheetjs.com/xlsx-0.19.3/xlsx-0.19.3.tgz"
},
"overrides": {
  "xlsx": "https://cdn.sheetjs.com/xlsx-0.19.3/xlsx-0.19.3.tgz"
}
```

That will override the `xlsx` package for your project, and you will now have all the negative consequences of relying on a third-party CDN.

**Comment by SymbioticKilla** [comment]
@Pauan 
It is not about whos fault is. We have bought two licenses. Now we get the information that you are relying on package that is not supported anymore (already a year) and you see no problem.

**Comment by sassomedia** [comment]
@SymbioticKilla I think they have explained very clearly what the underlying circumstances are, why it is not a security issue in AmChart's use case, why it does not need to be resolved on AmChart's side, and what other possible workaround there is.

I don't know what rights you think you have bought with your licenses but as any reasonable person would understand, no software comes with any sort of guarantee that it will work indefinitely or for every imaginable use case. You are free to choose another software that works with your policies.

**Comment by martynasma** [comment]
> @Pauan It is not about whos fault is. We have bought two licenses. Now we get the information that you are relying on package that is not supported anymore (already a year) and you see no problem.

As a representative of amCharts, I do hear you. In an ideal world I would like for all packages we and the rest of community rely on to be indefinitely supported and developed with the best of intentions.

However, the real situation is different, especially in the OS world. If you would dig deeper through other dependencies and their sub-sub-sub-dependencies, you'd find dinosaurs that are way older than the version of xlsx lib we are using :)

That said, this vulnerability does not affect amCharts functionality (we don't read xlsx files) and there are no known bugs in the older version that would cause amCharts-related functionality not to function correctly. If any of the above would cease to be the case, we'd go to great lengths to somehow amend it.

I hope this makes sense, and we can put this to rest. This is a space to discuss technical issues, not engage in discussions. If you'd still like to continue this discussion with us directly, as a client, please do contact us directly at contact@amcharts.com

Thank you for understanding.

**Comment by Pauan** [comment]
> @SymbioticKilla It is not about whos fault is. We have bought two licenses. Now we get the information that you are relying on package that is not supported anymore (already a year) and you see no problem.

That's an incorrect way of describing the situation. We do see a problem with it, but there is nothing we can do about it.

XLSX has chosen to stop uploading new versions to npm. We cannot force them to publish new versions to npm. It is their code, not our code. It is their project, not our project. We have no control over it.

Even if we were held at gunpoint we cannot fix the issue, it is not our repository. We do not have access to their npm account, so we cannot publish new versions on their behalf. So you will have to ask them to publish a new version.

Every company and organization uses third party code, that is standard practice. Unfortunately that means sometimes those third party projects can behave poorly. We don't really have the resources to create our own XLSX implementation, which is why we are relying on a third party in the first place.

Also, although it is admirable that your organization has a zero-vulnerability rule, that rule is misguided when it comes to npm, because [npm audit is fundamentally flawed](https://overreacted.io/npm-audit-broken-by-design/). That particular vulnerability with XLSX is not actually a vulnerability, it is a false positive.

In my previous message I gave a workaround which will allow you to use the latest version of XLSX (but with some caveats). That should allow you to fulfill your company's zero-vulnerability rule.

**Comment by jlgonzalezmindden** [comment]
The vulnerability is scaled to high. Are they not going to apply another alternative solution internally or other xlsx library?

![image](https://user-images.githubusercontent.com/47416384/235884391-2680a7ee-060b-4b80-981b-8e617c72948f.png)


Thanks in advance.

**Comment by martynasma** [comment]
> ... other xlsx library?

 @jlgonzalezmindden any suggestions?

**Comment by martynasma** [comment]
> There's a new related discussion at: https://git.sheetjs.com/sheetjs/sheetjs/issues/2935

Based on their reply, they're not budging :(

Seems like we'll need to do something on our end.

Stay tuned.

**Comment by jackwagons** [comment]
This issue is tied to v5 but will need to understand the path forward for AmCharts v4 as well.

**Comment by martynasma** [comment]
Heads up, we're cloning and bundling 0.19.3 into next release. Amazing how people can make it so much harder to use their software because of some principles.

Anyways, will notify here when the new version is out.

**Comment by martynasma** [comment]
> This issue is tied to v5 but will need to understand the path forward for AmCharts v4 as well.

Would you mind opening an issue in [v4's repo](https://github.com/amcharts/amcharts4/)?

**Comment by martynasma** [comment]
Fixed in [5.3.12](https://github.com/amcharts/amcharts5/releases/tag/5.3.12).

## [5.3.12] - 2023-05-05

### Added
- New locale: Faroese (fo_FO). Thanks Martin Puge.

### Changed
- The `min` and `max` settings are now optional for `IHeatRule` ([Issue 883](https://github.com/amcharts/amcharts5/issues/883)).
- Removed `xlsx` package (with vulnerability) from dependencies in favor of a bundled hard copy (version 0.19.3) due to their decision not to publish to NPM. ([Issue 897](https://github.com/amcharts/amcharts5/issues/897)).

### Fixed
- Events now work properly with open shadow DOM (but not closed shadow DOM).
- Fixed export of data freezing in some cases where there are visual elements on the chart that are constantly being updated.
- Using parentheses in the in-line number formatting function was not working as needed.
- Tapping on a touch display outside chart was applying default state to elements that weren't being "touch-hovered".
- `getDataItemById()` method was not working properly on `Hierarchy` series.

[Full change log](https://github.com/amcharts/amcharts5/blob/master/packages/shared/CHANGELOG.md).

[Download options](https://www.amcharts.com/download/).

Make sure you clear your browser cache after upgrading. And feel free to contact us again if you are still experiencing this issue.
